### PR TITLE
Add unit tests 🧪

### PR DESCRIPTION
**Pull request by last-genius**
Best reviewed by commit - I'm adding the basic skeleton for unit tests first, then tests themselves group by group. These are also added to the CI.

I've tried to minimize the amount of code mocked (as that might need some updating in the future), so there's some trickery with `Testing_status` (new in this PR) and `geteuid` (this was there in Edwin's branch too) rather than fragile copying of giant Xenbus, Xenmmap files.

The unit tests are pretty thorough, but if anyone has some edge case they want to add as well, please let me know.

### CODE REVIEW

**Comment by last-genius** [comment]
(You can't see that tests are passing in the CI in the main repo yet, but my fork is green: https://github.com/last-genius/oxenstored/tree/private/asultanov/unittests)

File: xenbus/xb.mli [hunk 1]
```diff
@@ -137,6 +137,8 @@ val has_output : t -> bool
 
 val peek_output : t -> Packet.t
 
+val unsafe_pop_output : t -> Packet.t
```
Comment (psafont) [inline xenbus/xb.mli:140]: The binding in the mli doesn't look any different from the other bindings, when it should not be used in nornmal conditions. One way to make it stand out while reading the mli would be to put it at the end of the file, below a comment explaining the rest of exposed bindings is for testing only. A way to make it stand out while using autocompletion would be add a doc-comment with a warning that it must be used only for testing.

**Review by psafont — APPROVED** [review]
This es excellent. I din't read the added tests in details. For me they're about documenting current behaviour so if the tests pass, they're good enough.

File: README.md [hunk 2]
```diff
@@ -104,3 +107,4 @@ The full list of commands is available in [process.ml](https://github.com/xapi-p
 ## Known incompatibilities with C Xenstore
 
 * chmod for dead domains will work with O, but not with C
+* Oxenstored does not preserve the order of nodes added to a directory (see `_test_directory_order` in [`unit_tests.ml`](./tests/unit_tests.ml))
```
Comment (lindig) [inline README.md:110]: Would suggest to wrap lines in Markdown files

File: gnt/gnttab_stubs.c [hunk 3]
```diff
@@ -73,26 +85,29 @@ stub_gnttab_unmap (value xgh, value array)
...
-			  caml_failwith(s);
-		}
-	}
+        xengnttab_handle *xgt = _G (xgh);
+        void *start_address = _M (array)->addr;
+        uint32_t count = _M (array)->len >> XEN_PAGE_SHIFT;
+        char s[64];
+
+        /* Check if this grant hasn't already been unmapped before */
+        if (start_address)
```
Comment (lindig) [inline gnt/gnttab_stubs.c:94]: I notice different formatting styles - has this been run through `make format`

File: oxenstored/domains.ml [hunk 4]
```diff
@@ -35,14 +35,17 @@ let load_plug fname =
...
-  let filepath = plugins_dir ^ "domain_getinfo_v1.cmxs" in
-  debug "Trying to load plugin '%s'\n%!" filepath ;
-  let list_files = Sys.readdir plugins_dir in
-  debug "Directory listing of '%s'\n%!" plugins_dir ;
-  Array.iter (fun x -> debug "\t%s\n%!" x) list_files ;
-  Dynlink.allow_only ["Plugin_interface_v1"] ;
-  load_plug filepath
+  (* Do not link with the plugin when being tested *)
+  if not Testing_status.under_testing then (
+    let plugins_dir = Paths.libexec ^ "/ocaml/xsd_glue/xenctrl_plugin/" in
```
Comment (lindig) [inline oxenstored/domains.ml:40]: Can we use `Filename.concat`?

File: xenbus/xb.mli [hunk 5]
```diff
@@ -145,3 +146,8 @@ val is_selectable : t -> bool
 val get_fd : t -> Unix.file_descr
 
 val debug : t -> string
+
+(* WARNING: Functions below this point are to be used in testing only *)
+
```
Comment (lindig) [inline xenbus/xb.mli:151]: There is a convention to hide such functions after 
```
(**/**)
```

in the MLI file. See `strings.mli` in OCaml as an example.

**Review by lindig — APPROVED** [review]
Excellent

File: gnt/gnttab_stubs.c (hunk 3 above)
Comment (last-genius) [reply gnt/gnttab_stubs.c:94]: yes - in the first commit
